### PR TITLE
Saving interactions

### PR DESCRIPTION
**Pull request by fang-helen**
Added a system of saving users' interactions with events as an "Interaction" entity in datastore, with a userId, eventId, rating (depending on the type of interaction) and timestamp. Only 1 interaction is saved for each userId-event-Id pair (saving the highest rating). Additionally, user preferences are updated according to the same rating system and the event's tags.

I also did a lot of refactoring (which is also why lines-changed is so high), where I moved some methods around and also put the user entity creation into its own Utils method.

### CODE REVIEW

File: src/main/java/com/google/sps/Interactions.java [hunk 1]
```diff
@@ -93,4 +105,70 @@ public static int dotProduct(Map<String, Integer> v1, Map<String, Integer> v2) {
...
+      interactionEntity = pq.asSingleEntity();
+      int prevScore = Integer.parseInt(interactionEntity.getProperty("rating").toString());
+      if (forceOverride || prevScore < score) {
+        interactionEntity.setProperty("rating", score);
+        delta = score - prevScore;
+      } else {
+        delta = 0;
+      }
+    } catch (TooManyResultsException e) {
+      // delete existing entries and re-create later
```
Comment (Yunni) [inline src/main/java/com/google/sps/Interactions.java:140]: Please log whenever an exception is caught.

File: src/main/java/com/google/sps/Recommend.java [hunk 2]
```diff
@@ -20,10 +27,24 @@ private static void init() {
...
   /** Rebuilds recommendation model and calculates recommendations for users. */
-  public static void calculateRecommend() {}
+  public static void calculateRecommend() {
+    if (spark == null || datastore == null) {
+      init();
+    }
+    Iterable<Entity> queriedUsers = datastore.prepare(new Query("User")).asIterable();
+    // get user prefs
+    Map<String, Map<String, Integer>> userPrefs = new HashMap<>();
+    for (Entity e : queriedUsers) {
```
Comment (Yunni) [inline src/main/java/com/google/sps/Recommend.java:42]: For debugging?

File: src/main/java/com/google/sps/Utils.java [hunk 3]
```diff
@@ -144,21 +173,21 @@ public int compare(Entity a, Entity b) {
...
-            }
-            Map<String, Integer> v1 = Interactions.buildVectorForEvent(a);
-            Map<String, Integer> v2 = Interactions.buildVectorForEvent(b);
-            return Integer.compare(
-                Interactions.dotProduct(v1, metrics), Interactions.dotProduct(v2, metrics));
-          }
-        };
-    return result;
-  }
+  //   /** creates a comparator based on entity dot product with a given vector. (not used) */
```
Comment (Yunni) [inline src/main/java/com/google/sps/Utils.java:176]: Please remove commented code in production. You can recover it from git history if you need it later.

File: src/main/java/com/google/sps/Recommend.java (hunk 2 above)
Comment (fang-helen) [reply src/main/java/com/google/sps/Recommend.java:42]: That's just a method stub, I haven't gotten around to implementing anything yet.

File: src/main/java/com/google/sps/Utils.java (hunk 3 above)
Comment (fang-helen) [reply src/main/java/com/google/sps/Utils.java:176]: Got it!

File: src/main/java/com/google/sps/Interactions.java (hunk 1 above)
Comment (fang-helen) [reply src/main/java/com/google/sps/Interactions.java:140]: Got it!

**Pull request merged**